### PR TITLE
Proposal/classmap

### PR DESCRIPTION
**Pull request by chfsx**
A static classmap is faster than scanning the filesystem for existing classes structured by PSR-4.
As an addition I included all ILIAS-Classes form Services and Modules to the Classmap (without */lib). No require_once needed from now on!

Since the composer classmap can be stored to APC-cache this is blazing fast.

Only painpoint i see: we have to update the classmap from time to time

### CODE REVIEW

File: libs/composer/composer.json [hunk 1]
```diff
@@ -29,7 +32,9 @@
 		"mockery/mockery": "^0.9.9"
 	},
 	"autoload": {
-		"psr-4" : { "ILIAS\\" : "../../src" }
+		"psr-4" : { "ILIAS\\" : "../../src" },
+		"classmap": ["../../Services/", "../../Modules"],
+		"exclude-from-classmap": ["../../Services/Migration", "../../*/*/lib", "../../*/*/test", "../../Services/CAS/phpcas"]
```
Comment (chfsx) [inline libs/composer/composer.json:40]: @smeyer-ilias Something a bit of-topic: I found an external library in Services/CAS/phpcas which is not in the subdir Services/CAS/libs/... . The phpcas also could be added using composer. If you have the time, would be cool to move it

**Comment by chfsx** [comment]
There are currently a few ambigous classes:

```
Warning: Ambiguous class resolution, "ilLDAPAccountMigrationRequiredException" was found in both "/var/www/ilias/Services/LDAP/exceptions/class.ilLDAPAccountMigrationRequiredException.php" and "/var/www/ilias/Services/LDAP/exceptions/class.ilLDAPConnectTimeOutException.php", the first will be used.
Warning: Ambiguous class resolution, "ilContObjectManifestBuilder" was found in both "/var/www/ilias/Modules/LearningModule/classes/class.ilContObjectManifestBuilder.php" and "/var/www/ilias/Modules/Scorm2004/classes/class.ilContObjectManifestBuilder.php", the first will be used.
Warning: Ambiguous class resolution, "assLongMenuExport" was found in both "/var/www/ilias/Modules/TestQuestionPool/classes/export/qti12/class.assLongMenuExport.php" and "/var/www/ilias/Modules/TestQuestionPool/classes/export/qti21/class.assLongMenuExport.php", the first will be used.
```

**Comment by gvollbach** [comment]
`assLongMenuExport` fixed, since this is not really used anywhere because ILIAS doesn't support QTI 2.1 Export/Import yet, I changed the name for this class.

**Review by klees — CHANGES_REQUESTED** [review]
Hi @chfsx,

in general I like this (speed...), but the pain point you mentioned to me seems to be rather severe. 

We will at least need a description how to update the class map, because "from time to time" is not enough when I just added a class during development. But this only is the bare minimum.

This might also be a case for the DEV-mode, one could use the psr-4 autoloader during development and the classmap for production.

I would suggest to step back a littleand think about this autoloading thing in general. There are already issues with the dev-dependencies that cause unwanted changes in my repo, and I see similar problems arise due to this change. I think there are systematic ways to deal with both problems (git-hooks or proper build process) but I fear that developers experience get worse if we merge this PR now.

Best regards!

**Comment by chfsx** [comment]
Actually the problem isn't that severe: The class-map is updated whenever a developer adds new or updates dependencies using composer. Therefore all classes are available for him after the changes in composer.
The ILIAS classes in the classmap can be used (I for my part will kick out most `requires` from my code) if wanted. The proposal here doesn't forbid to use `requires`, but nothing bad happens when you forget one. I for my part would update the classmap (`composer dump-autoload`) instead of putting a new `require_once` into my code. But as already said, I do not force developers the go this way with this proposal.

I agree that an automatic process for the generation of the classmap would be nice but I think this can be done in a second step. Additionally I will add a short description to the readme.

**Review by klees — APPROVED** [review]

**Review by alex40724 — APPROVED** [review]
I like it. First I thought that we might get lots of conflicts when all developers do this, but these will be exceptional cases, since new entries are added alphabetically and not simply to the end.

**Comment by matthiaskunkel** [comment]
Jour Fixe, June 19, 2017: Pull request is accepted. Please merge to trunk and notify developers over the developers list and in the developer forum about this issue.

**Pull request merged**